# This patch 'fixes' the GRPC build system so it can be used on
# openwrt for cross-compiling. It is based on the buildroot patch:
# https://github.com/buildroot/buildroot/blob/5099a1058d93821399ea758c5227e1fbbe6f76de/package/grpc/0003-host-grpc-only-cpp-plugin.patch
# that allows you to build  just the CPP plugin. It has been modified to also:
#
# 1. Allow you to build the CPP _and_ Python plugin
# 2. Fix the cmake/protobuf.cmake file to not use cmake's find_program function
#    if we provide a manual override via a variable.
#
# These two modifications are necessary to be able to compile the
# right programs for the host and also to use the host's protoc
# compiler instead of the version compiled for the target.

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4d9a95c918..b6564d7ab7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -49,12 +49,16 @@ option(gRPC_BUILD_TESTS "Build tests" OFF)
 option(gRPC_BUILD_CODEGEN "Build codegen" ON)
 option(gRPC_BUILD_CSHARP_EXT "Build C# extensions" ON)
 option(gRPC_BACKWARDS_COMPATIBILITY_MODE "Build libraries that are binary compatible across a larger number of OS and libc versions" OFF)
+option(gRPC_BUILD_PLUGIN_SUPPORT_ONLY "Build plugin support only" OFF)
 
 set(gRPC_INSTALL_default ON)
 if(NOT CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
   # Disable gRPC_INSTALL by default if building as a submodule
   set(gRPC_INSTALL_default OFF)
 endif()
+if(gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
+  set(gRPC_INSTALL_default OFF)
+endif()
 set(gRPC_INSTALL ${gRPC_INSTALL_default} CACHE BOOL
     "Generate installation target")
 
@@ -557,6 +561,8 @@ add_custom_target(plugins
   DEPENDS ${_gRPC_PLUGIN_LIST}
 )
 
+if (NOT gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
+
 add_custom_target(tools_c
   DEPENDS
 )
@@ -569,6 +575,8 @@ add_custom_target(tools_cxx
 add_custom_target(tools
   DEPENDS tools_c tools_cxx)
 
+endif() # gRPC_BUILD_PLUGIN_SUPPORT_ONLY
+
 protobuf_generate_grpc_cpp_with_import_path_correction(
   src/proto/grpc/channelz/channelz.proto src/proto/grpc/channelz/channelz.proto
 )
@@ -1280,6 +1288,7 @@ if(gRPC_BUILD_TESTS)
     DEPENDS buildtests_c buildtests_cxx)
 endif()
 
+if (NOT gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
 
 add_library(address_sorting
   third_party/address_sorting/address_sorting.c
@@ -1332,6 +1341,8 @@ if(gRPC_INSTALL)
   )
 endif()
 
+endif() # gRPC_BUILD_PLUGIN_SUPPORT_ONLY
+
 if(gRPC_BUILD_TESTS)
 
 add_library(end2end_tests
@@ -1507,6 +1518,8 @@ endforeach()
 
 endif()
 
+if(NOT gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
+
 add_library(gpr
   src/core/lib/gpr/alloc.cc
   src/core/lib/gpr/atm.cc
@@ -2480,6 +2493,8 @@ if(gRPC_INSTALL)
   )
 endif()
 
+endif() # gRPC_BUILD_PLUGIN_SUPPORT_ONLY
+
 if(gRPC_BUILD_TESTS)
 
 add_library(grpc_test_util
@@ -2594,6 +2609,8 @@ endif()
 
 endif()
 
+if(NOT gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
+
 add_library(grpc_unsecure
   src/core/ext/filters/census/grpc_context.cc
   src/core/ext/filters/channel_idle/channel_idle_filter.cc
@@ -3041,6 +3058,8 @@ if(gRPC_INSTALL)
   )
 endif()
 
+endif() # gRPC_BUILD_PLUGIN_SUPPORT_ONLY
+
 if(gRPC_BUILD_TESTS)
 
 if(gRPC_BUILD_CODEGEN)
@@ -3124,6 +3143,8 @@ endif()
 
 endif()
 
+if(NOT gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
+
 add_library(grpc++
   src/core/ext/transport/binder/client/binder_connector.cc
   src/core/ext/transport/binder/client/channel_create.cc
@@ -3570,6 +3591,7 @@ if(gRPC_INSTALL)
   )
 endif()
 
+endif() # gRPC_BUILD_PLUGIN_SUPPORT_ONLY
 
 if(gRPC_BUILD_CODEGEN)
 add_library(grpc++_reflection
@@ -3824,6 +3846,8 @@ target_link_libraries(grpc++_test_util
 
 endif()
 
+if(NOT gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
+
 add_library(grpc++_unsecure
   src/cpp/client/channel_cc.cc
   src/cpp/client/client_callback.cc
@@ -4111,6 +4135,7 @@ if(gRPC_INSTALL)
   )
 endif()
 
+endif() # gRPC_BUILD_PLUGIN_SUPPORT_ONLY
 
 add_library(grpc_plugin_support
   src/compiler/cpp_generator.cc
@@ -4170,7 +4195,7 @@ foreach(_hdr
 endforeach()
 
 
-if(gRPC_INSTALL)
+if(gRPC_INSTALL OR gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
   install(TARGETS grpc_plugin_support EXPORT gRPCTargets
     RUNTIME DESTINATION ${gRPC_INSTALL_BINDIR}
     BUNDLE DESTINATION  ${gRPC_INSTALL_BINDIR}
@@ -4254,6 +4279,8 @@ endif()
 
 endif()
 
+if(NOT gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
+
 add_library(upb
   third_party/upb/third_party/utf8_range/naive.c
   third_party/upb/third_party/utf8_range/range2-neon.c
@@ -4351,6 +4378,8 @@ target_link_libraries(gen_hpack_tables
   gpr
 )
 
+endif() # gRPC_BUILD_PLUGIN_SUPPORT_ONLY
+
 
 if(gRPC_BUILD_TESTS)
 
@@ -10859,7 +10888,7 @@ target_link_libraries(grpc_completion_queue_test
 
 
 endif()
-if(gRPC_BUILD_CODEGEN AND gRPC_BUILD_GRPC_CPP_PLUGIN)
+if(gRPC_BUILD_GRPC_CPP_PLUGIN)
 
 add_executable(grpc_cpp_plugin
   src/compiler/cpp_plugin.cc
@@ -10889,7 +10918,7 @@ target_link_libraries(grpc_cpp_plugin
 
 
 
-if(gRPC_INSTALL)
+if(gRPC_INSTALL OR gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
   install(TARGETS grpc_cpp_plugin EXPORT gRPCTargets
     RUNTIME DESTINATION ${gRPC_INSTALL_BINDIR}
     BUNDLE DESTINATION  ${gRPC_INSTALL_BINDIR}
@@ -11106,7 +11135,8 @@ if(gRPC_INSTALL)
 endif()
 
 endif()
-if(gRPC_BUILD_CODEGEN AND gRPC_BUILD_GRPC_PYTHON_PLUGIN)
+
+if(gRPC_BUILD_GRPC_PYTHON_PLUGIN)
 
 add_executable(grpc_python_plugin
   src/compiler/python_plugin.cc
@@ -21882,7 +21912,7 @@ endif()
 
 
 
-
+if (NOT gRPC_BUILD_PLUGIN_SUPPORT_ONLY)
 
 if(gRPC_INSTALL)
   install(EXPORT gRPCTargets
@@ -21979,3 +22009,5 @@ generate_pkgconfig(
   "-lgrpc++_unsecure"
   ""
   "grpc++_unsecure.pc")
+
+endif() # gRPC_BUILD_PLUGIN_SUPPORT_ONLY
diff --git a/cmake/protobuf.cmake b/cmake/protobuf.cmake
index f23f65d40e..a6c05f455f 100644
--- a/cmake/protobuf.cmake
+++ b/cmake/protobuf.cmake
@@ -41,6 +41,8 @@ if(gRPC_PROTOBUF_PROVIDER STREQUAL "module")
       set(_gRPC_PROTOBUF_PROTOC protoc)
       if(CMAKE_CROSSCOMPILING)
         find_program(_gRPC_PROTOBUF_PROTOC_EXECUTABLE protoc)
+	message(STATUS "Finding protoc with find_program, result (overriding): ${_gRPC_PROTOBUF_PROTOC_EXECUTABLE}")
+	set(_gRPC_PROTOBUF_PROTOC_EXECUTABLE /home/jphillips/code-projects/openwrt/staging_dir/hostpkg/bin/protoc)
       else()
         set(_gRPC_PROTOBUF_PROTOC_EXECUTABLE $<TARGET_FILE:protoc>)
       endif()
@@ -56,7 +58,6 @@ if(gRPC_PROTOBUF_PROVIDER STREQUAL "module")
   endif()
 elseif(gRPC_PROTOBUF_PROVIDER STREQUAL "package")
   find_package(Protobuf REQUIRED ${gRPC_PROTOBUF_PACKAGE_TYPE})
-
   # {Protobuf,PROTOBUF}_FOUND is defined based on find_package type ("MODULE" vs "CONFIG").
   # For "MODULE", the case has also changed between cmake 3.5 and 3.6.
   # We use the legacy uppercase version for *_LIBRARIES AND *_INCLUDE_DIRS variables
@@ -75,7 +76,10 @@ elseif(gRPC_PROTOBUF_PROVIDER STREQUAL "package")
       set(_gRPC_PROTOBUF_PROTOC_LIBRARIES ${PROTOBUF_PROTOC_LIBRARIES})
       set(_gRPC_PROTOBUF_WELLKNOWN_INCLUDE_DIR ${PROTOBUF_INCLUDE_DIRS})
     endif()
-    if(TARGET protobuf::protoc)
+    if(gRPC_PROTOBUF_PROTOC_EXECUTABLE)
+	message(STATUS "Overriding protoc to: ${gRPC_PROTOBUF_PROTOC_EXECUTABLE}")
+      	set(_gRPC_PROTOBUF_PROTOC_EXECUTABLE "${gRPC_PROTOBUF_PROTOC_EXECUTABLE}")
+    elseif(TARGET protobuf::protoc)
       set(_gRPC_PROTOBUF_PROTOC protobuf::protoc)
       if(CMAKE_CROSSCOMPILING)
         find_program(_gRPC_PROTOBUF_PROTOC_EXECUTABLE protoc)
